closed/src/java.base/share/classes/com/sun/crypto/provider/NativeChaCha20Cipher.java

```diff
@@ -935,12 +936,49 @@ int doFinal(byte[] in, int inOff, int inLen, byte[] out, int outOff)
         * @return the number of unprocessed bytes left.
         */
         int getCipherBufferLength();
+
+        /**
+         * Determines if two arrays have any overlapping memory.
+         *
+         * @param input The input array.
+         * @param inputStart The index of the input arrays start.
+         * @param inputEnd The index of the input arrays end.
+         * @param output The output array.
+         * @param outputStart The index of the output arrays start.
+         * @param outputEnd The index of the output arrays end.
+         * @return true if any memory overlaps, false otherwise.
+         */
+        static boolean arraysOverlap(byte[] input, int inputStart, int inputEnd, byte[] output, int outputStart, int outputEnd) {
+
+            // Check if there is potential overlap if input and output buffers have same address.
+            if (input != output) {
+                return false;
+            }
+
+            // If the start of input is anywhere in the range of the output array, there is an overlap.
+            if (outputStart <= inputStart && inputStart <= outputEnd) {
+                return true;
+            }
+
+            // If the start of output is anywhere in the range of the input array, there is an overlap.
+            if (inputStart <= outputStart && outputStart <= inputEnd) {
```

Comment:
Please add parentheses and `<` should be used instead of `<=` for the end part of the conditions:
```java
            if ((outputStart <= inputStart) && (inputStart < outputEnd)) {
```

Reply:
I added parentheses in latest code.

Reply:
I think this method would not work as intended if we changed the comparison to `<` instead of `<=`. The case I have in mind is the following:
- Same buffer being used
- inputStart = 1
- inputEnd = 4
- outputStart = 4
- outputEnd = 8
We want this to return true to indicate an overlay with this set of parameters to this method.

In this case the inputEnd and outputStart overlap by just a single byte, I think we would need `<=` to catch this condition at the end of line 964.

Reply:
No, based on how this function is used, the end offsets are exclusive. So if input is `[1,4)` its length is 3 and only bytes at indices 1, 2 and 3 are accessed which doesn't overlap with an output beginning at offset 4 or larger.

Reply:
OK yes does seem like this is the case in how this method is used. I updated the code using commit [f7f939c](https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/744/commits/f7f939c52246361b747e1bdae5f32e515e753c66) to reflect this.